src/js/index.d.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * A list of routes and their parameters and bindings.
+ *
+ * Extended and filled by the route list generated automatically by Ziggy.
+ */
+export interface RouteList {
+    //
+}
+
+/**
+ * A route name registered with Ziggy.
+ */
+type KnownRouteName = keyof RouteList;
+
+/**
+ * A route name, or any string.
+ */
+type RouteName = KnownRouteName | (string & {});
+
+/**
+ * Information about a single route parameter.
+ */
+type ParameterInfo = { name: string, binding?: string };
+
+/**
+ * A primitive route parameter value.
+ */
+type ParameterValue = string | number;
+
+/**
+ * A route parameter value, possibly nested inside an object under its binding key.
+ */
+type ValueOrBoundValue<I extends ParameterInfo> = I extends { binding: string }
+    ? { [K in I['binding']]: ParameterValue } | ParameterValue
+    : ParameterValue
+// type A = ValueOrBoundValue<{ name: 'foo', binding: 'bar' }>;
+// = ParameterValue | { bar: ParameterValue }
+
+/**
+ * An object containing a special '_query' key to target the query string of a URL.
+ */
+type HasQueryParam = { _query?: Record<string, any> };
+/**
+ * An object of parameters for an unknown route.
+ */
+type GenericRouteParamsObject = Record<keyof any, any> & HasQueryParam;
+/**
+ * An object of parameters for a specific named route registered with Ziggy.
+ */
+type KnownRouteParamsObject<I extends readonly ParameterInfo[]> = { [T in I[number] as T['name']]?: ValueOrBoundValue<T> | string } & HasQueryParam;
```

Comment:
What is `readonly` doing in `I extends readonly ParameterInfo[]>`?

How exactly does `T in I[number] as T['name']` work? I can _kind of_ wrap my head around what it ends up accomplishing, but the syntax isn't clicking for me. Does `I[number]` enable a second level of iteration through `I`? Why not just `T in I`?

Why `| string`? In my testing everything seems to work the same without it, and `ValueOrBoundValue<T>` already includes `string`.

Reply:
> What is readonly doing in I extends readonly ParameterInfo[]>?

Taking `ParameterInfo[]` as `readonly` makes TS consider it as `const` array. I.e. there is a difference between
```typescript
const parameterInfo: ParameterInfo[] = [{ ...paramInfo1, name: 'pi1' }, { ...paramInfo2, name: 'pi2' }]
const roParameterInfo: ParameterInfo[] = [{ ...paramInfo1, name: 'pi1' }, { ...paramInfo2, name: 'pi2' }] as const
```

The expressions allow typescript to resolves name the exact string literals, whereas the first one resolves to string. So: 

```typescript
type Names = (typeof parameterInfo)[number]['name'] // String  but
type RoNames = (typeof roParamterInfo)[number]['name'] // 'pi1' | 'pi2'
```

Disclaimer: I am not sure if these snippets exactly reproduce in TS, but this is the basic principle behind that.

Reply:
> How exactly does T in I[number] as T['name'] work? I can kind of wrap my head around what it ends up accomplishing, but the syntax isn't clicking for me. Does I[number] enable a second level of iteration through I? Why not just T in I?

Given that
```typescript
type I = readonly Array<{ name: string }>
```

Then we can map it to an object where each key is the name of each array element. E.g refine the expression above to a more concrete:

```typescript
const arr = [{ name: 'foo' }, { name: 'bar' }] as const;
type Q = typeof arr;
```

Now we create an object from that array where the key is the name property of all elements:

```typescript
type ObjType = {  [K in keyof Q[number] as K['name']]: SomeType }
//                 ^          ^^^^^^^^^    ^^^^^^^^^
//                 |          |            The actual key name, e.g. 'foo' | 'bar'
//                            The "iteratee" is not the array itself (since `keyof []` is push, pop, splice, ....) but the elements
//                 The "iteration" variable which is basically the type of each element in the array
```

Thus, the type of `ObjType` from the concrete `Q` is mapped to
```typescript
type ObjTypeOfQ = {
  foo: SomeType
  bar: SomeType
} 
```

Reply:
> Why | string? In my testing everything seems to work the same without it, and ValueOrBoundValue<T> already includes string.

Probably a mistake. I did a lot of experiments, created new types, dropped them, moved some parts of the unions around. Probably just a relict of that.

Reply:
Very cool, that makes sense and is super helpful. Thanks a lot!